src/amp-story-embed-manager.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpStoryEmbed} from './amp-story-embed';
+
+/** @const {string} */
+const SCROLL_THROTTLE_MS = 500;
+
+export class AmpStoryEmbedManager {
+  /**
+   * @param {!Window} win
+   * @constructor
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+  }
+
+  /**
+   * Calls layoutCallback on the embed when it is close to the viewport.
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutEmbed_(embedImpl) {
+    if (!IntersectionObserver || this.win_ !== this.win_.parent) {
+      this.layoutFallback_(embedImpl);
+      return;
+    }
+
+    const intersectingCallback = entries => {
+      entries.forEach(entry => {
+        if (!entry.isIntersecting) {
+          return;
+        }
+        embedImpl.layoutCallback();
+      });
+    };
+
+    const observer = new IntersectionObserver(intersectingCallback, {
+      rootMargin: '100%',
+    });
+    observer.observe(embedImpl.getElement());
+  }
+
+  /**
+   * Fallback for when IntersectionObserver is not supported. Calls
+   * layoutCallback on the embed when it is close to the viewport.
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutFallback_(embedImpl) {
+    let tick = true;
+
+    // TODO(Enriqe): pause embeds when scrolling away from viewport.
+    this.win_.addEventListener('scroll', () => {
```

Comment:
You need to unregister this eventually.

Reply:
We will need it later to pause the component when the user scrolls away from the component.

Reply:
You're not using it to pause now. Is this being added in a later PR?

Reply:
Correct, this will come in a follow up PR. 👍